### ISSUE TITLE
[AVM Module Issue]: Not able to set Customer-managed key for CosmosDBAccount

### ISSUE THREAD

**Issue by dmusic** [Azure/bicep-registry-modules#3748]
### Check for previous/existing GitHub issues

- [x] I have checked for previous/existing GitHub issues

### Issue Type?

Feature Request

### Module Name

avm/res/document-db/database-account

### (Optional) Module Version

0.8.1

### Description

> Description

In CosmosDB Account AVM, it is not possible to set customer-managed key for encryption.

Following properties are missing:

- keyVaultKeyUri - The URI of the key vault
- defaultIdentity - The default identity for accessing key vault used in features like customer managed keys. The default identity needs to be explicitly set by the users. It can be "FirstPartyIdentity", "SystemAssignedIdentity" and more.

See plain Bicep: https://learn.microsoft.com/en-us/azure/templates/Microsoft.DocumentDB/2023-04-15/databaseAccounts?pivots=deployment-language-bicep

This would enable direct encryption of CosmosDB Account with CMK.

**Comment by seesharprun** [comment]
This should be a supported scenario.

@bryansan-msft, would you be okay with me taking an attempt at adding this feature?

**Comment by seesharprun** [comment]
Here's the relevant documentation: https://learn.microsoft.com/en-us/azure/cosmos-db/how-to-setup-cross-tenant-customer-managed-keys?tabs=azure-cli#using-azure-resource-manager-json-templates

I could see something like:

```bicep
keyVault: {
  uri: '<string>'
  identityConfiguration: '<string>'
}
```

So an example configuration would be:

```bicep
keyVault: {
  uri: 'https://my-vault.vault.azure.com/keys/my-key'
  identityConfiguration: 'UserAssignedIdentity=/subscriptions/aaaa0a0a-bb1b-cc2c-dd3d-eeeeee4e4e4e/resourcegroups/my-resource-group/providers/Microsoft.ManagedIdentity/userAssignedIdentities/my-identity'
}
```

I like the name `keyVault.identityConfiguration` better than `defaultIdentity` because it's less vague and specifies this is a configuration string used to configure the identity specifically for use with Azure Key Vault. I think `keyVault.Uri` is self-documenting enough.

**Comment by dmusic** [comment]
I agree naming seems not to be conform with others CMK product implementations so it would be great to clean it up and wrap accordingly.
This causes an issue if you want to use Continuous backup policy as this policy needs user-managed identity to be used and not default parity-identity: https://learn.microsoft.com/en-us/azure/cosmos-db/how-to-setup-customer-managed-keys?tabs=arm-template#use-customer-managed-keys-with-continuous-backup

I think implementation should be straight forward.

**Comment by seesharprun** [comment]
@bryansan-msft, would you be okay with me taking an attempt at adding this feature?

**Comment by dmusic** [comment]
any news on this one?

**Comment by mark-vw** [comment]
Any updates on this one @bryansan-msft

**Comment by mark-vw** [comment]
The customerManagedKeyType is available in the avm-common-types and is used widely among other AVM modules. @seesharprun / @bryansan-msft is there a timeline for getting this issue fixed?

**Comment by dmusic** [comment]
I have to admit I am a bit disappointed with AVM operating model. I understand that it is dependent on availability and that it does not have any SLAs. However, Microsoft has promoted it and recommend its usage with regular improvements and maintenance. I doubt now in longevity of the approach and its future. 
To give just an example: I've recently was tasked with automation using Azure Data Factory and its module has missed again specific features. At the end, I've decided to use vanilla Bicep and implement it myself. 
AVM are great idea but it lacks on executions and resources as it seems. I think in future, I will continue using plain/vanilla Bicep (unfortunately :( ).

**Comment by AlexanderSehr** [comment]
Hey folks,
while 'only' a general maintainer and not the owner of the module, let me take a stab at this issue. I know that @seesharprun had to continuously repriotize, which makes addressing issues challenging even over a period of time. It's easier if one has one or multiple co-owners for a module, but it is what it is.
@dmusic, you're in good company and as far as I'm concerned pivoting to whatever suits your needs best is the right cause of action. Others may not share the same sentiment, but at least my take is that even if one does not use an AVM module, the library can still be a great reference for 'how' to implement certain features & their dependencies. 💪 Regardless, I of course hope that we can continuously work towards to stable and complete operations model and eventually get there. 🤞

Back to the issue at hand: I'll let you know and link a PR once I got anywhere. As @mark-vw suggested I'll look into implementing the  `customerManagedKeyType`, though it might be a custom version of it as the CosmosDB PG seemed to have implemented a unique interface for the feature.

**Comment by dmusic** [comment]
Hello @AlexanderSehr, 

I see a new property "keyVaultKeyUri" in your pull request: https://github.com/Azure/bicep-registry-modules/pull/5958/commits/b4bf3e0d436f7abfe70b01e2c4a4a4d0d219720e

Is there CMK feature coming in into module finally? Do you have a timeline for this?

Thanks for info.

**Comment by AlexanderSehr** [comment]
Hey @dmusic,
that's the plan, right. Actually it was the plan to already have this merged 10 days ago as the implementation is fairly simple. But the property you called out (`keyVaultKeyUri`) is driving me a bit nuts as the Resource Provider just seems to ignore it even though the REST API & docs all point to it.
I'm since then hence in a troubleshooting mode, trying to figure out what's happening and why the RP would ignore this property. I can't give you an exact timeline on this as this should be a non-issue and I'm trying to find pick around as long as I need to in order to find out where the issue is burried. I can however tell you that I'm looking into it as often as I can as I want to get this done.

PS: If you have any wild guesses regarding the issue or see an obvious flaw, let me know 😄

**Comment by AlexanderSehr** [comment]
Hey @dmusic, please note that I just created a support ticket as I was unable to get the IaC deployment to accept the configuration & even when trying the UI I got some suspicious behavior as can be seen [here](https://github.com/Azure/bicep-registry-modules/pull/5958#issuecomment-3325184793) (<= tried to set the same key twice in a row, first failing, then working).

**Comment by ghost** [comment]
**@dmusic, thanks for submitting this issue for the `avm/res/document-db/database-account` module!**

> [!IMPORTANT]
> Please note, that this module is currently orphaned. The @Azure/avm-core-team-technical-bicep, will attempt to find an owner for it. In the meantime, the core team may assist with this issue. Thank you for your patience!

**Comment by AlexanderSehr** [comment]
I case anybody is wondering - yes, I'm still very much still chasing this issue down. As it's about engaging with a set of engineering teams it's a very async communication though. 
What is (not) happening is pretty obvious and we've been able to narrow it down quite a bit. But the why is still in the air.

**Comment by Thaomoeba** [comment]
@AlexanderSehr throwing in a message to say that this is a need for our org as well.

**Comment by AlexanderSehr** [comment]
Ok now finally a good update: We finally managed to find out what the issue ended up being, and it's not something that will require an update to the resource provider, but rather the documentation. Turns out, the (failover) `locations` parameter **must** be configured for the CMK settings to be applied. Why? Who knows - and I hope this is either changed, or at least documented. I wouldn't know how one would guess that.
Anyways, I updated the module accordingly in the linked PR, and the `locations` parameter will always be populated (by default with the resource's location). In the current version, the property would only be set if a database is also deployed to the resource. As I'm not very knowledgeable with the resource I can't say why it was implemented this way, but I won't question it either. Anyways - I'm running a few final tests and then will ready the PR.

**Comment by AlexanderSehr** [comment]
Version `0.19.0`  is now published and ready for use. You can find an example of how to use the feature [here](https://github.com/Azure/bicep-registry-modules/tree/main/avm/res/document-db/database-account#example-5-using-encryption-with-customer-managed-key).
